File: ai_diagnostic/alerts.py
"""
Alert and notification system for diagnostic results.

Provides rule-based alerting with configurable actions including
webhooks, email, logging, and console output.
"""

import json
import smtplib
import sys
import urllib.request
from dataclasses import dataclass, field
from datetime import datetime, timezone
from email.message import EmailMessage
from typing import Callable, List, Optional

from .bayesian import DiagnosisResult
from .likelihoods import FAILURE_CLASSES


@dataclass
class AlertRule:
    name: str
    condition: Callable[[DiagnosisResult], bool]
    actions: List[Callable[[DiagnosisResult, str], None]]


class ThresholdAlert:
    """Helper for confidence/posterior threshold checks."""

    def __init__(self, threshold: float, direction: str = "above"):
        if direction not in ("above", "below"):
            raise ValueError(f"direction must be 'above' or 'below', got {direction!r}")
        if not (0.0 <= threshold <= 1.0):
            raise ValueError(f"threshold must be in [0, 1], got {threshold}")
        self.threshold = threshold
        self.direction = direction

    def check_confidence(self, result: DiagnosisResult) -> bool:
        if self.direction == "above":
            return result.confidence > self.threshold
        return result.confidence < self.threshold

    def check_posterior(self, cause: str, result: DiagnosisResult) -> bool:
        if cause not in FAILURE_CLASSES:
            raise ValueError(f"Unknown cause: {cause}")
        prob = result.posteriors.get(cause, 0.0)
        if self.direction == "above":
            return prob > self.threshold
        return prob < self.threshold


class AlertManager:
    """Rule-based alert manager for diagnosis results."""

    def __init__(self):
        self._rules: List[AlertRule] = []
        self._triggered: List[str] = []

    def add_rule(
        self,
        name: str,
        condition: Callable[[DiagnosisResult], bool],
        actions: List[Callable[[DiagnosisResult, str], None]],
    ) -> None:
        self._rules.append(AlertRule(name=name, condition=condition, actions=actions))

    def check(self, result: DiagnosisResult) -> List[str]:
        self._triggered.clear()
        for rule in self._rules:
            if rule.condition(result):
                self._triggered.append(rule.name)
                for action in rule.actions:
                    action(result, rule.name)
        return list(self._triggered)

    def add_webhook(
        self, name: str, url: str, method: str = "POST"
    ) -> Callable[[DiagnosisResult, str], None]:
        def webhook_action(result: DiagnosisResult, rule_name: str) -> None:
            payload = json.dumps({
                "rule": rule_name,
                "timestamp": datetime.now(timezone.utc).isoformat(),
                **result.to_dict(),
            }).encode("utf-8")
            req = urllib.request.Request(
                url,
                data=payload,
                headers={"Content-Type": "application/json"},
                method=method,
            )
            try:
                urllib.request.urlopen(req, timeout=10)
            except Exception as exc:
                print(
                    f"[AlertManager] webhook {name!r} failed: {exc}",
                    file=sys.stderr,
                )

        return webhook_action

    @staticmethod
    def add_log_action(path: str) -> Callable[[DiagnosisResult, str], None]:
        def log_action(result: DiagnosisResult, rule_name: str) -> None:
            entry = {
                "timestamp": datetime.now(timezone.utc).isoformat(),
                "rule": rule_name,
                **result.to_dict(),
            }
            with open(path, "a") as f:
                f.write(json.dumps(entry) + "\n")

        return log_action

    @staticmethod
    def add_console_action() -> Callable[[DiagnosisResult, str], None]:
        def console_action(result: DiagnosisResult, rule_name: str) -> None:
            label = FAILURE_CLASSES[
                list(FAILURE_CLASSES).index(result.diagnosis)
            ] if result.diagnosis in FAILURE_CLASSES else result.diagnosis
            print(
                f"[ALERT] {rule_name}: diagnosis={result.diagnosis} "
                f"confidence={result.confidence:.2%}",
                file=sys.stderr,
            )

        return console_action

    @staticmethod
    def add_email_action(
        smtp_host: str, to: str, from_addr: str
    ) -> Callable[[DiagnosisResult, str], None]:
        def email_action(result: DiagnosisResult, rule_name: str) -> None:
            msg = EmailMessage()
            msg["Subject"] = f"[AI Diagnostic Alert] {rule_name}"
            msg["From"] = from_addr
            msg["To"] = to
            body = (
                f"Alert triggered: {rule_name}\n"
                f"Diagnosis: {result.diagnosis}\n"
                f"Confidence: {result.confidence:.2%}\n"
                f"Needs investigation: {result.needs_investigation}\n\n"
                f"Posterior probabilities:\n"
            )
            for cause, prob in sorted(
                result.posteriors.items(), key=lambda x: x[1], reverse=True
            ):
                body += f"  {cause}: {prob:.4f}\n"
            msg.set_content(body)
            try:
                with smtplib.SMTP(smtp_host, timeout=10) as smtp:
                    smtp.send_message(msg)
            except Exception as exc:
                print(
                    f"[AlertManager] email to {to} failed: {exc}",
                    file=sys.stderr,
                )

        return email_action

    def preset_rules(self) -> List[AlertRule]:
        presets = [
            AlertRule(
                name="critical_malice",
                condition=lambda r: r.diagnosis == "human_malice" and r.confidence > 0.9,
                actions=[self.add_console_action()],
            ),
            AlertRule(
                name="warning_low_confidence",
                condition=lambda r: r.confidence < 0.5,
                actions=[self.add_console_action()],
            ),
            AlertRule(
                name="entropy_alert",
                condition=lambda r: r.diagnosis == "entropy" and r.confidence > 0.8,
                actions=[self.add_console_action()],
            ),
            AlertRule(
                name="investigation_needed",
                condition=lambda r: r.needs_investigation,
                actions=[self.add_console_action()],
            ),
        ]
        self._rules.extend(presets)
        return presets

File: ai_diagnostic_rs/src/alerts.rs
//! Rule-based alerting with webhook, email, log-file, and console actions.

use crate::bayesian::DiagnosisResult;
use crate::likelihoods::{failure_class_label, FAILURE_CLASSES};
use serde_json::json;

#[derive(Debug, Clone)]
pub enum Direction {
    Above,
    Below,
}

/// Confidence/posterior threshold checks.
#[derive(Debug, Clone)]
pub struct ThresholdAlert {
    pub threshold: f64,
    pub direction: Direction,
}

impl ThresholdAlert {
    pub fn new(threshold: f64, direction: Direction) -> Result<Self, String> {
        if !(0.0..=1.0).contains(&threshold) {
            return Err(format!("threshold must be in [0, 1], got {threshold}"));
        }
        Ok(ThresholdAlert {
            threshold,
            direction,
        })
    }

    pub fn check_confidence(&self, result: &DiagnosisResult) -> bool {
        match self.direction {
            Direction::Above => result.confidence > self.threshold,
            Direction::Below => result.confidence < self.threshold,
        }
    }

    pub fn check_posterior(&self, cause: &str, result: &DiagnosisResult) -> Result<bool, String> {
        if !FAILURE_CLASSES.contains(&cause) {
            return Err(format!("Unknown cause: {cause}"));
        }
        let prob = result.posterior(cause);
        Ok(match self.direction {
            Direction::Above => prob > self.threshold,
            Direction::Below => prob < self.threshold,
        })
    }
}

pub type Condition = Box<dyn Fn(&DiagnosisResult) -> bool>;
pub type Action = Box<dyn Fn(&DiagnosisResult, &str)>;

pub struct AlertRule {
    pub name: String,
    pub condition: Condition,
    pub actions: Vec<Action>,
}

pub struct AlertManager {
    rules: Vec<AlertRule>,
    triggered: Vec<String>,
}

impl Default for AlertManager {
    fn default() -> Self {
        Self::new()
    }
}

impl AlertManager {
    pub fn new() -> Self {
        AlertManager {
            rules: Vec::new(),
            triggered: Vec::new(),
        }
    }

    pub fn add_rule(
        &mut self,
        name: impl Into<String>,
        condition: Condition,
        actions: Vec<Action>,
    ) {
        self.rules.push(AlertRule {
            name: name.into(),
            condition,
            actions,
        });
    }

    /// Check all rules against a result; fire actions of matching rules.
    /// Returns the names of triggered rules.
    pub fn check(&mut self, result: &DiagnosisResult) -> Vec<String> {
        self.triggered.clear();
        let rules = &mut self.rules;
        for rule in rules.iter_mut() {
            if (rule.condition)(result) {
                self.triggered.push(rule.name.clone());
                for action in &rule.actions {
                    action(result, &rule.name);
                }
            }
        }
        self.triggered.clone()
    }

    pub fn webhook_action(url: &str, method: &str) -> Action {
        let url = url.to_string();
        let method = method.to_string();
        Box::new(move |result: &DiagnosisResult, rule_name: &str| {
            let payload = json!({
                "rule": rule_name,
                "timestamp": chrono::Utc::now().to_rfc3339(),
                "posteriors": result
                    .posteriors
                    .iter()
                    .map(|(k, v)| (k.clone(), json!(v)))
                    .collect::<serde_json::Map<String, serde_json::Value>>(),
                "diagnosis": result.diagnosis,
                "diagnosis_label": failure_class_label(&result.diagnosis),
                "confidence": result.confidence,
                "needs_investigation": result.needs_investigation,
                "log_likelihoods": result
                    .log_likelihoods
                    .iter()
                    .map(|(k, v)| (k.clone(), json!(v)))
                    .collect::<serde_json::Map<String, serde_json::Value>>(),
            });
            match method.to_uppercase().as_str() {
                "POST" => ureq::post(&url)
                    .timeout(std::time::Duration::from_secs(10))
                    .send_json(payload),
                _ => ureq::request(&method.to_uppercase(), &url)
                    .timeout(std::time::Duration::from_secs(10))
                    .send_json(payload),
            }
            .map(|_| ())
            .unwrap_or_else(|e| {
                eprintln!("[AlertManager] webhook {rule_name:?} failed: {e}");
            });
        })
    }

    pub fn log_action(path: &str) -> Action {
        use std::io::Write;
        let path = path.to_string();
        Box::new(move |result: &DiagnosisResult, rule_name: &str| {
            let entry = json!({
                "timestamp": chrono::Utc::now().to_rfc3339(),
                "rule": rule_name,
                "posteriors": result
                    .posteriors
                    .iter()
                    .map(|(k, v)| (k.clone(), json!(v)))
                    .collect::<serde_json::Map<String, serde_json::Value>>(),
                "diagnosis": result.diagnosis,
                "diagnosis_label": failure_class_label(&result.diagnosis),
                "confidence": result.confidence,
                "needs_investigation": result.needs_investigation,
                "log_likelihoods": result
                    .log_likelihoods
                    .iter()
                    .map(|(k, v)| (k.clone(), json!(v)))
                    .collect::<serde_json::Map<String, serde_json::Value>>(),
            });
            if let Ok(mut f) = std::fs::OpenOptions::new()
                .create(true)
                .append(true)
                .open(&path)
            {
                let _ = writeln!(f, "{entry}");
            }
        })
    }

    pub fn console_action() -> Action {
        Box::new(|result: &DiagnosisResult, rule_name: &str| {
            eprintln!(
                "[ALERT] {rule_name}: diagnosis={} confidence={:.2}%",
                result.diagnosis,
                result.confidence * 100.0
            );
        })
    }

    pub fn email_action(smtp_host: &str, to: &str, from_addr: &str) -> Action {
        use lettre::message::{Mailbox, Message};
        use lettre::{Address, SmtpTransport, Transport};

        let host = smtp_host.to_string();
        let to = to.to_string();
        let from_addr = from_addr.to_string();

        Box::new(move |result: &DiagnosisResult, rule_name: &str| {
            let run = || -> Result<(), String> {
                // smtp_host may include a port ("host:port"); default to 25 like smtplib.SMTP.
                let (host_part, port) = match host.rsplit_once(':') {
                    Some((h, p)) if p.parse::<u16>().is_ok() => {
                        (h.to_string(), p.parse::<u16>().unwrap())
                    }
                    _ => (host.clone(), 25),
                };

                let to_address: Address = to.parse().map_err(|_| format!("invalid To address: {to}"))?;
                let from_address: Address = from_addr
                    .parse()
                    .map_err(|_| format!("invalid From address: {from_addr}"))?;

                let mut body = format!(
                    "Alert triggered: {rule_name}\n\
                     Diagnosis: {}\n\
                     Confidence: {:.2}%\n\
                     Needs investigation: {}\n\n\
                     Posterior probabilities:\n",
                    result.diagnosis,
                    result.confidence * 100.0,
                    result.needs_investigation
                );
                let mut sorted = result.posteriors.clone();
                sorted.sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap_or(std::cmp::Ordering::Equal));
                for (cause, prob) in &sorted {
                    body.push_str(&format!("  {cause}: {prob:.4}\n"));
                }

                let email = Message::builder()
                    .from(Mailbox::new(None, from_address))
                    .to(Mailbox::new(None, to_address))
                    .subject(format!("[AI Diagnostic Alert] {rule_name}"))
                    .body(body)
                    .map_err(|e| e.to_string())?;

                let mailer = SmtpTransport::builder_dangerous(&host_part)
                    .port(port)
                    .timeout(Some(std::time::Duration::from_secs(10)))
                    .build();
                mailer.send(&email).map_err(|e| e.to_string())?;
                Ok(())
            };
            if let Err(e) = run() {
                eprintln!("[AlertManager] email to {to} failed: {e}");
            }
        })
    }

    /// The four built-in rules.
    pub fn preset_rules(&mut self) -> Vec<String> {
        let presets: Vec<(&str, Condition)> = vec![
            (
                "critical_malice",
                Box::new(|r: &DiagnosisResult| r.diagnosis == "human_malice" && r.confidence > 0.9),
            ),
            (
                "warning_low_confidence",
                Box::new(|r: &DiagnosisResult| r.confidence < 0.5),
            ),
            (
                "entropy_alert",
                Box::new(|r: &DiagnosisResult| r.diagnosis == "entropy" && r.confidence > 0.8),
            ),
            (
                "investigation_needed",
                Box::new(|r: &DiagnosisResult| r.needs_investigation),
            ),
        ];
        let mut names = Vec::new();
        for (name, condition) in presets {
            self.add_rule(name, condition, vec![Self::console_action()]);
            names.push(name.to_string());
        }
        names
    }
}
